File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */
func createDummyNode(head *Node){
    temp:=head
    for temp!=nil{
        copyNode:=&Node{Val:temp.Val}
        copyNode.Next=temp.Next
        temp.Next=copyNode
        temp=temp.Next.Next
    }
}
func fixRandomPtr(head *Node){
    temp:=head;
    for temp!=nil{
        copyNode:=temp.Next
        if(temp.Random!=nil){
            copyNode.Random=temp.Random.Next
        }else{
            copyNode.Random=nil;
        }
        temp=temp.Next.Next
    }
}
func fixNextPtr(head *Node)* Node{
    dummyNode:=&Node{Val:-1}
    res:=dummyNode
    temp:=head
    for temp!=nil{
        res.Next=temp.Next;
        temp.Next=temp.Next.Next
        
        res=res.Next;
        temp=temp.Next
    }
    return dummyNode.Next;
}
func copyRandomList(head *Node) *Node {
    createDummyNode(head)
    fixRandomPtr(head)
    return fixNextPtr(head)
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    void createDummyNode(Node* head){
        Node* temp=head;
        while(temp){
            Node *copyNode = new Node(temp->val);
            copyNode->next=temp->next;
            temp->next=copyNode;
            temp=temp->next->next;
        }
    }
    void connectRandomPtr(Node* head){
        Node* temp=head;
        while(temp){
            Node* copyNode=temp->next;
            if(temp->random){
                copyNode->random=temp->random->next;
            }else{
                copyNode->random=NULL;
            }
            temp=temp->next->next;
        }
    }
    Node* connectNextPtr(Node *head){
        Node* dummyNode=new Node(-1);
        Node* res=dummyNode;
        Node* temp=head;
        while(temp){
            res->next=temp->next;
            temp->next=temp->next->next;
            
            res=res->next;
            temp=temp->next;
        }
        return dummyNode->next;
    }
    Node* copyRandomList(Node* head) {
        createDummyNode(head);
        connectRandomPtr(head);
        return connectNextPtr(head);
    }
};